"url" field is undefined when restricting returned fields via the API

vkandy opened the pull request:
Closes #6625 

`GET /posts` query with `?field=url` as query string returns "undefined" as url value. The `url` property is constructed from id, published_dt, and slug. Post instances are loaded with attributes given in `fields` parameter. If the `fields` parameter doesn't contain `id`, `published_dt`, and `slug`, URL can't be constructed and undefined is returned.

Fix was to load the 3 required properties when `field` parameter contains `url`. Response is sent after filtering out any properties that were not requested.

- "url" and "author" fields depend on {id, published_at, slug, author_id} to construct post url.
- results are transformed by filtering out additional properties to return just the requested fields

vkandy commented:
Please review.

Response returns `author_id` when `?field=author` is passed. Do we return name or id or both? The docs aren't clear: http://api.ghost.org/docs/fields

ErisDS commented:
The author/author_id aspect is a slightly blurry boundary because it is a related object.

Posts have a `field` called `author_id` and a related object called `author`. Usually control over this is provided by using `?include=author` to include the full related object instead of the id. 

Ideally at this point, you would be able to say `include=author&fields=author.name` in order to get just an author name, but handling joins in fields is a separate issue.

ErisDS commented:
I have tested this out with the reproduction steps [listed in this comment](https://github.com/TryGhost/Ghost/issues/6625#issuecomment-232625892) and am no longer able to reproduce either bug 😉

vkandy commented:
Added a test case for this fix. Good for review.

kirrg001 commented:
i was able to reproduce with postman. will review

kirrg001 commented on core/server/api/posts.js:
```diff
@@ -55,7 +55,27 @@ posts = {
+            if (_.isArray(options.columns)) {
+                requestedColumns = _.clone(options.columns);
+                if (_.includes(options.columns, 'url') || _.includes(options.columns, 'author')) {
+                    options.columns.push('id', 'published_at', 'slug', 'author_id');
```
instead of defining a specific solution here, i would suggest making this part of the model logic. something like pseudo:

```
fetchPage
  if (options.columns) {
     // only allow permitted columns
     options.columns = _.intersection(options.columns, this.prototype.permittedAttributes());

     // ensure minimum required columns for further logic
     options.columns union this.defaultColumnsToFetch();
  }

// per model (can be empty)
defaultColumnsToFetch: function()
   return 'id, slug....';
```

vkandy replied:
Just to make sure we are on the same page... I think we only need these 4 extra properties when `url` or `author` is present in the `fields` parameter. 
Do we treat them as default columns to fetch if at least 1 column is requested?

kirrg001 replied:
there are two ways: either we fetch by default almost everything and pick what the user has requested for (maybe a performance issue) or we always fetch a minimum set of attributes per model (depends what each model needs minimum to fulfil its needs).

in this case: there is a somewhere a piece of logic in the `toJSON` fn to generate a dynamic URL, which depends on a set of attributes. in my opinion we shouldn't add a another piece of logic somewhere in a controller, so that the `toJSON` will not return undefined.

vkandy commented:
Based on @kirrg001 suggestion above, implemented a more generic solution. Base class for models defines `defaultColumnsToFetch()` which is overridden by subclasses. In this case, Posts overrides this method and returns `['id', 'published_at', 'slug', 'author_id']`.

`findPage()` loads model instances by performing union of permitted columns and default columns declared by models. 

Finally, the response from `toJSON()` is filtered and only columns requested by the client are returned.

kirrg001 commented on core/server/models/base/index.js:
```diff
@@ -355,14 +362,21 @@ ghostBookshelf.Model = ghostBookshelf.Model.extend({
 
         return itemCollection.fetchPage(options).then(function formatResponse(response) {
-            var data = {};
+            options.columns = allColumns;
```
suggestion:
now that we extend `options.columns`, maybe we can rename `allColumns` to `requestedColumns` (or something similar)

kirrg001 commented on core/server/models/base/index.js:
```diff
@@ -355,14 +362,21 @@ ghostBookshelf.Model = ghostBookshelf.Model.extend({
-            data[tableName] = response.collection.toJSON(options);
-            data.meta = {pagination: response.pagination};
+            // pick only requested before returning JSON
+            return Promise.map(models, function (model) {
```
suggestion: 
no need to use `Promise.map`, here is no async operation happening. 
```
// pick only requested ...
models = _.map(models, function(model) {...});
```

ErisDS commented:
The failing lint build was a random travis error - this is green now!

vkandy commented:
As per the suggestions above, renamed `allColumns` to `requestedColumns` (these are the attributes client asked for).
Replaced `Promise.map()` with `_.map()`. My rationale was that we could benefit from concurrency on large arrays but I guess the savings are minimal and `_.map` is more readable.
